src/expr/src/relation.rs

```diff
@@ -3544,3 +3551,298 @@ mod tests {
+                            input: input2,
+                        },
+                    ) => {
+                        if outputs1 != outputs2 {
+                            return Some((expr1, expr2));
+                        } else {
+                            self.todo.push((input1, input2));
+                        }
+                    }
+                    (
+                        MirRelationExpr::Map {
+                            scalars: scalars1,
+                            input: input1,
+                        },
+                        MirRelationExpr::Map {
+                            scalars: scalars2,
+                            input: input2,
+                        },
+                    ) => {
+                        if scalars1 != scalars2 {
+                            return Some((expr1, expr2));
+                        } else {
+                            self.todo.push((input1, input2));
+                        }
+                    }
+                    (
+                        MirRelationExpr::Filter {
+                            predicates: predicates1,
+                            input: input1,
+                        },
+                        MirRelationExpr::Filter {
+                            predicates: predicates2,
+                            input: input2,
+                        },
+                    ) => {
+                        if predicates1 != predicates2 {
+                            return Some((expr1, expr2));
+                        } else {
+                            self.todo.push((input1, input2));
+                        }
+                    }
+                    (
+                        MirRelationExpr::FlatMap {
+                            input: input1,
+                            func: func1,
+                            exprs: exprs1,
+                        },
+                        MirRelationExpr::FlatMap {
+                            input: input2,
+                            func: func2,
+                            exprs: exprs2,
+                        },
+                    ) => {
+                        if func1 != func2 || exprs1 != exprs2 {
+                            return Some((expr1, expr2));
+                        } else {
+                            self.todo.push((input1, input2));
+                        }
+                    }
+                    (
+                        MirRelationExpr::Join {
+                            inputs: inputs1,
+                            equivalences: eq1,
+                            implementation: impl1,
+                        },
+                        MirRelationExpr::Join {
+                            inputs: inputs2,
+                            equivalences: eq2,
+                            implementation: impl2,
+                        },
+                    ) => {
+                        if eq1 != eq2 || impl1 != impl2 {
```

Comment:
Also check `inputs.len`.

Reply:
Smart! I think I got this right in `Union` and didn't revisit `Join`.